playground/engine.py

```diff
@@ -0,0 +1,818 @@
+        reflectance_scatter = mesh.faces.new_zeros(size=(num_faces,))
+        refractive_index = Primitives.DEFAULT_REFRACTIVE_INDEX
+        refractive_index_tensor = mesh.faces.new_full(size=(num_faces,), fill_value=refractive_index)
+
+        self.objects[name] = OptixPrimitive(
+            geometry_type=geometry_type,
+            vertices=mesh.vertices.float(),
+            triangles=mesh.faces.int(),
+            vertex_normals=mesh.vertex_normals.float(),
+            has_tangents=has_tangents.bool(),
+            vertex_tangents=mesh.vertex_tangents.float(),
+            material_uv=mesh.face_uvs.float(),
+            material_id=mesh.material_assignments.unsqueeze(1).int(),
+            primitive_type=primitive_type,
+            primitive_type_tensor=prim_type_tensor.int(),
+            reflectance_scatter=reflectance_scatter.float(),
+            refractive_index=refractive_index,
+            refractive_index_tensor=refractive_index_tensor.float(),
+            transform=transform
+        )
+
+    def remove_primitive(self, name: str):
+        del self.objects[name]
+        self.rebuild_bvh_if_needed(True, True)
+
+    def duplicate_primitive(self, name: str):
+        prim = self.objects[name]
+        geometry_type = prim.geometry_type
+        self.instance_counter[prim.geometry_type] += 1
+        name = f"{geometry_type} {self.instance_counter[geometry_type]}"
+        self.objects[name] = copy.deepcopy(prim)
+        self.rebuild_bvh_if_needed(True, True)
+
+    def register_materials(self, materials, model_name: str):
+        """ Registers list of material dictionaries.
+        """
+        mat_idx_to_mat_id = torch.full([len(materials)], -1)
+        for mat_idx, mat in enumerate(materials):
+            material_name = f'{model_name}${mat["material_name"]}'
+            if material_name not in self.registered_materials:
+                self.registered_materials[material_name] = PBRMaterial(
+                    material_id=len(self.registered_materials),
+                    diffuse_map=mat['diffuse_map'],
+                    emissive_map=mat['emissive_map'],
+                    metallic_roughness_map=mat['metallic_roughness_map'],
+                    normal_map=mat['normal_map'],
+                    diffuse_factor=mat['diffuse_factor'],
+                    emissive_factor=mat['emissive_factor'],
+                    metallic_factor=mat['metallic_factor'],
+                    roughness_factor=mat['roughness_factor'],
+                    alpha_mode=mat['alpha_mode'],
+                    alpha_cutoff=mat['alpha_cutoff'],
+                    transmission_factor=mat['transmission_factor'],
+                    ior=mat['ior']
+                )
+            mat_idx_to_mat_id[mat_idx] = self.registered_materials[material_name].material_id
+        return mat_idx_to_mat_id
+
+    def create_geometry(self, geometry_type: str, device):
+        match geometry_type:
```

Comment:
`match` / `case` are Python >=3.10. Not sure what our minimum version is, but this makes it at least 3.10.